### PR TITLE
fixed bug with undefined this.placeholder in allowClear.js and update…

### PR DESCRIPTION
**Pull request by joshi1983**
…d corresponding output files

This bug and fix was mentioned in #3817

This pull request includes a

- [x] Bug fix
- [ ] New feature
- [ ] Translation

The following changes were made

Like the commit says, I applied a fix mentioned in the comments of that issue.  It fixed problems for me and I wanted to share the fix with any other select2 user.

-[https://github.com/select2/select2/issues/3917](https://github.com/select2/select2/issues/3917)

If this is related to an existing ticket, include a link to it as well.

### CODE REVIEW

**Comment by jpic** [comment]
Nice patch ! Can you add a test case ? :guitar:

File: dist/js/select2.js [hunk 1]
```diff
@@ -1802,7 +1802,11 @@ S2.define('select2/selection/allowClear',[
       }
     }
 
-    this.$element.val(this.placeholder.id).trigger('change');
+    if (typeof this.placeholder !== 'undefined') {
+        this.$element.val(this.placeholder.id).trigger('change');
+    } else {
+        this.$element.val('').trigger('change');
+    }
```
Comment (jpic) [inline dist/js/select2.js]: Please remove changes in dist: https://github.com/select2/select2/blob/master/CONTRIBUTING.md#submitting-a-pull-request

File: dist/js/select2.full.js [hunk 2]
```diff
@@ -1802,7 +1802,11 @@ S2.define('select2/selection/allowClear',[
       }
     }
 
-    this.$element.val(this.placeholder.id).trigger('change');
+    if (typeof this.placeholder !== 'undefined') {
+        this.$element.val(this.placeholder.id).trigger('change');
+    } else {
+        this.$element.val('').trigger('change');
+    }
```
Comment (jpic) [inline dist/js/select2.full.js]: Change in dist/ better left off: https://github.com/select2/select2/blob/master/CONTRIBUTING.md#submitting-a-pull-request

File: dist/js/select2.js (hunk 1 above)
Comment (joshi1983) [inline dist/js/select2.js]: Why is the distribution tracked when it is generated off other sources?

Comment (jpic) [inline dist/js/select2.js]: We don't want it to create conflicts with other PRs, it'll be updated later. Looking forward to review the test case :guitar:

File: dist/js/select2.full.js (hunk 2 above)
Comment (joshi1983) [inline dist/js/select2.full.js]: dist directory changes have been removed.

**Comment by joshi1983** [comment]
I removed the changes to the dist directory but creating a test for the bug in #3917 might take a while for me because I'd have to dig deeper into this bug and how the tests work while being pretty busy.  I tried to trace the code and tests/selection/allowClear-tests.js seems like a good place for one but it could be a couple weeks.  If that's too long, can someone else build take my commit and finish off the test case?

**Comment by jpic** [comment]
No problem, I'd rather wait two months now, than see a regression in two months and have you redo your patch ^^

**Comment by hxnan** [comment]
The same error I met. 
I can not find this fix in the master now(2017-11-25 18:13).
Is this fix has't merge to the master or another fix ?

**Comment by joshi1983** [comment]
@hxnan it isn't in master yet.

You could make the requested test and submit a new pull request that is more merge-able to help this fix get into master.

**Comment by joshi1983** [comment]
Hopefully someone can finish my fix for this so the issue doesn't get closed by the github bot.  It still fixes the bug even though there's no automated test on it.

**Pull request closed without merging**